src/themes/JSONTheme.ts

```diff
@@ -0,0 +1,1003 @@
+        shadow_ctx: CanvasRenderingContext2D | null,
+        stone: any,
+        cx: number,
+        cy: number,
+        radius: number,
+    ) {
+        if (this.blackImages.length > 0) {
+            if (shadow_ctx && this.blackShadowImages.length > 0) {
+                const img = this.blackShadowImages[stone.rando % this.blackShadowImages.length];
+
+                const t = shadow_ctx.getTransform();
+
+                shadow_ctx.translate(cx, cy);
+                shadow_ctx.scale(radius * 2.0, radius * 2.0);
+
+                const m = this.matrices[stone.rando % this.matrices.length]["blackShadowMatrix"];
+                shadow_ctx.transform(...m);
+
+                shadow_ctx.drawImage(img, -0.5, -0.5, 1.0, 1.0); // unit box centered around cx, cy
+
+                shadow_ctx.setTransform(t);
+            }
+
+            const img = this.blackImages[stone.rando % this.blackImages.length];
+
+            const t = ctx.getTransform();
+            ctx.translate(cx, cy);
+            ctx.scale(radius * 2.0, radius * 2.0);
+            const m = this.matrices[stone.rando % this.matrices.length]["blackMatrix"];
+            ctx.transform(...m);
+
+            ctx.drawImage(img, -0.5, -0.5, 1.0, 1.0); // unit box centered around cx, cy
+
+            ctx.setTransform(t);
+        } else {
+            if (shadow_ctx && this.blackShadowImages.length > 0) {
+                const img = this.blackShadowImages[stone.rando % this.blackShadowImages.length];
+
+                const t = shadow_ctx.getTransform();
+                shadow_ctx.translate(cx, cy);
+                shadow_ctx.scale(radius * 2.0, radius * 2.0);
+                const m = this.matrices[stone.rando % this.matrices.length]["blackShadowMatrix"];
+                shadow_ctx.transform(...m);
+                shadow_ctx.drawImage(img, -0.5, -0.5, 1.0, 1.0); // unit box centered around cx, cy
+
+                shadow_ctx.setTransform(t);
+            }
+            const t = ctx.getTransform();
+
+            ctx.save();
+            ctx.translate(cx, cy);
+            ctx.scale(radius * 2, radius * 2);
+
+            const m = this.matrices[stone.rando % this.matrices.length]["blackMatrix"];
+            ctx.transform(...m);
+
+            ctx.fillStyle = this.getBlackStoneColor();
+            ctx.strokeStyle = this.getBlackStoneLineColor();
+            ctx.lineWidth = this.getBlackStoneLineWidth();
+            ctx.beginPath();
+            ctx.arc(0, 0, 0.5, 0, 2 * Math.PI, true);
+            ctx.fill();
+            if (this.getBlackStoneLineWidth() > 0) {
+                ctx.stroke();
+            }
+            ctx.restore();
+            ctx.setTransform(t);
+        }
+    }
+
+    /* Should return true if you would like the shadow layer to be present. False
+     * speeds up rendering typically */
+    public stoneCastsShadow(radius: number): boolean {
+        return true; // themes will always be given a shadow layer
```

Comment:
Is this a functional necessity? This exists because when we render a lot of mini boards on lower power devices, it made a notable difference.

Reply:
No, not really if the json theme doesn't have shadows. I think I assumed it was only asked once so I set it true to cover all bases.

Fixed with [4d92fb2](https://github.com/online-go/goban/pull/63/commits/4d92fb2cc3c8f28eae21e5a6dc8eca03cc2cbea2)